File: roguedoom_classic/PhysicEntity.java
import javax.vecmath.*;
import com.sun.j3d.utils.geometry.*;
import com.sun.j3d.utils.universe.*;
import javax.media.j3d.*;
import Labyrinth;
public class PhysicEntity extends Behavior
{
	public Vector3d collisionpoint = new Vector3d();
	public Vector3d position= new Vector3d();	
	public Vector3d direction= new Vector3d();	
	public Vector3d directionVel= new Vector3d();	
	public Vector3d vecside= new Vector3d();	
	public Vector3d perpendicular = new Vector3d();
	public Vector3d normal= new Vector3d();
	public Vector3d velocity= new Vector3d();
	double impulse=0.0,friction=0.2;
	double angle=0.0,angfriction=0.025;
	double radius=1.0;
	boolean applyfriction= true;	
	boolean applyanglefriction= true;	
	boolean applynormal= false;
	protected Labyrinth m_Labyrinth;
	protected Cell m_cell;
	protected BranchGroup	br= new BranchGroup();
	
	
	PhysicEntity(Labyrinth Labyrinth )
	{
		m_Labyrinth = Labyrinth;
		///setEnable(false);
		BoundingSphere bounds = new BoundingSphere(new Point3d(0.0,0.0,0.0), 1000.0);
		setSchedulingBounds(bounds);		
		br.setCapability(BranchGroup.ALLOW_DETACH);
		br.addChild(this);
		
		
		
	}
	
	public void create()
	{
		m_Labyrinth.getUniverse().addBranchGraph(br);	
	}
	
	
	public void initialize()
	{
		////position= new Vector3d(0.0,0.0,0.0);	
		direction= new Vector3d(0.0,0.0,1.0);	
		directionVel= new Vector3d(direction);
		vecside= new Vector3d(-1.0,0.0,0.0);	
		perpendicular = new Vector3d(0.0,1.0,0.0);	
		normal= new Vector3d(0.0,0.0,0.0);
		velocity= new Vector3d(0.0,0.0,0.0);		
		impulse=0.0;
		friction=0.08;
		angle=0.0;
		angfriction=0.05;
		radius=1.0;
		applyfriction= true;
		applyanglefriction= true;	
		applynormal= false;			
		
		
		m_cell = m_Labyrinth.getCell(0,0,false);
		
		position.set(m_Labyrinth.getCellPosition(m_cell.m_row,
										m_cell.m_column, m_cell.m_upstairs));
		position.y= position.y + Labyrinth.CELLHEIGHT/2;												
		wakeupOn(new WakeupOnElapsedTime(10));
		
		
		
	}
	public void applyFriction()
	{
		double signus = 1.0;
		if(applyfriction==false) return;
		
		
		if(impulse!=0.0)
		{
			if(impulse<0) signus = -1.0;
			impulse= impulse-friction*signus;
			if(signus>0.0)
			{
				if(impulse<0.0)	impulse=0.0;
			}
			else
			{
				if(impulse>0.0)	impulse=0.0;
			}	
		}
	}
	
	public void applyAngFriction()
	{
		double signus = 1.0;
		if(applyanglefriction==false) return;
		
		if(angle!=0.0)
		{
			if(angle<0) signus = -1.0;		
			angle= angle-angfriction*signus;
			if(signus>0.0)
			{
				if(angle<0.0)	angle=0.0;
			}
			else
			{
				if(angle>0.0)	angle=0.0;
			}	
		}
	}
	
	public void calcRotation()
	{
		if(angle==0.0)
		{			
			return;
		}
		
		Transform3D t= new Transform3D();				
		Vector3d vd = new Vector3d();		
		t.rotY(angle); 		
		t.transform(direction);
		t.transform(directionVel);
		t.transform(vecside);
		t.transform(perpendicular);
		applyAngFriction();
	}

	
	public void calcVelocity()
	{
		if(impulse==0.0)
		{
			velocity.set(new Vector3d(0.0,0.0,0.0));
			return;
		}
		Vector3d vdir=new Vector3d(directionVel);	
		Vector3d vnormal=new Vector3d(normal);
		
		applyFriction();	
		vdir.scale(impulse);		
		
		/*if(applynormal)
		{
			
			Vector3d auxvector= new Vector3d();		
			Vector3d auxvector2= new Vector3d();		
			auxvector.cross(normal,vdir);
			if(auxvector.length()>0.001)
			{			
				auxvector2.cross(normal,auxvector);	
				
				auxvector2.normalize();
				auxvector2.scale(vdir.dot(auxvector2));
				vdir.set(auxvector2);
		}
			
		}*/		
		
		velocity.set(vdir);
		
	}

	public void updatePosition()
	{
		
		Vector3d newpos= new Vector3d();
		Cell cel1[]= new Cell[1];		
		calcVelocity();
		newpos.set(position);
		newpos.add(velocity);		
		applynormal =m_Labyrinth.crossLabyrinth(radius,position ,newpos,normal,collisionpoint,cel1);		
		position.set(newpos);			
		calcRotation();
		
		m_cell = cel1[0];
	}

	public void processStimulus(java.util.Enumeration criteria)
	{
		updatePosition();
		wakeupOn(new WakeupOnElapsedTime(10));
	}
		
}

File: PhysicEntity.js
import * as THREE from 'three';
import { Labyrinth } from './Labyrinth.js'; // For Labyrinth.CELLWIDTH etc.
import { Cell } from './Cell.js'; // For Cell types

export class PhysicEntity {
    constructor(labyrinth) {
        this.labyrinth = labyrinth;

        this.position = new THREE.Vector3();
        this.direction = new THREE.Vector3(0, 0, 1); // Forward Z
        this.directionVel = new THREE.Vector3().copy(this.direction); // Velocity direction
        this.perpendicular = new THREE.Vector3(0, 1, 0); // Up vector
        this.vecside = new THREE.Vector3().crossVectors(this.direction, this.perpendicular).normalize(); // Should be (-1,0,0) for (0,0,1) dir

        this.normal = new THREE.Vector3(); // Collision normal
        this.collisionPoint = new THREE.Vector3(); // Point of collision

        this.velocity = new THREE.Vector3();
        this.impulse = 0.0;
        this.friction = 0.08;
        this.angle = 0.0; // Current angular change to apply (delta)
        this.rotationAngle = 0.0; // Accumulated rotation (like in PlayerCamera)
        this.angfriction = 0.05;

        this.radius = 1.0; // Default collision radius
        this.applyfriction = true;
        this.applyanglefriction = true;
        this.applynormal = false; // True if collision occurred

        this.m_cell = null; // Current cell
        // Initialize m_cell based on initial position (e.g., if position is set later)
        // For now, let's assume it will be set by the derived class or after position is known.
        // this.updateCurrentCell(); // Needs position to be set first
    }

    updateCurrentCell() {
        if (!this.labyrinth) return;
        const col = Math.floor(this.position.x / Labyrinth.CELLWIDTH);
        const row = Math.floor(this.position.z / Labyrinth.CELLWIDTH);
        // Determine if upstairs based on Y position, needs careful handling of level heights
        // This is a simplified assumption for Y. A more robust way would be needed if levels overlap in XZ.
        const floorThickness = 0.2;
        const ceilingThickness = 0.2;
        const singleLevelTotalHeight = Labyrinth.CELLHEIGHT + floorThickness + ceilingThickness;
        const upstairs = this.position.y > singleLevelTotalHeight / 2 + floorThickness; // Approx.

        this.m_cell = this.labyrinth.getCell(row, col, upstairs);
        return this.m_cell;
    }

    applyFriction() {
        if (this.applyfriction) {
            this.impulse *= (1.0 - this.friction);
            if (Math.abs(this.impulse) < 0.001) { // Threshold to stop
                this.impulse = 0.0;
            }
        }
    }

    applyAngFriction() {
        if (this.applyanglefriction) {
            this.angle *= (1.0 - this.angfriction);
            if (Math.abs(this.angle) < 0.001) { // Threshold to stop
                this.angle = 0.0;
            }
        }
    }

    // calcRotation is more complex and tied to how 'angle' is used.
    // If 'angle' is a delta, it modifies 'rotationAngle' which then determines 'direction'.
    calcRotation() {
        this.rotationAngle += this.angle; // Apply delta 'angle' to accumulated 'rotationAngle'
        
        // Similar to PlayerCamera, update direction based on accumulated rotationAngle
        this.direction.x = Math.sin(this.rotationAngle);
        this.direction.z = Math.cos(this.rotationAngle);
        this.direction.y = 0; // Assuming 2D rotation for direction
        this.direction.normalize();

        // Update vecside based on new direction and fixed perpendicular (up)
        this.vecside.crossVectors(this.direction, this.perpendicular).normalize();
        
        // directionVel should typically follow direction
        this.directionVel.copy(this.direction);

        this.applyAngFriction(); // Apply friction to the delta 'angle'
    }


    calcVelocity() {
        // Velocity is impulse along directionVel
        this.velocity.copy(this.directionVel).multiplyScalar(this.impulse);
    }

    updatePosition(deltaTime) {
        this.calcVelocity();

        let newpos = this.position.clone().add(this.velocity.clone().multiplyScalar(deltaTime));
        let endCellArray = [null]; // Array to hold the output cell

        if (this.labyrinth) {
            const collided = this.labyrinth.crossLabyrinth(
                this.radius,
                this.position, // current position
                newpos,       // desired new position (will be modified by crossLabyrinth on collision)
                this.normal,  // output normal (will be modified)
                this.collisionPoint, // output collision point (will be modified)
                endCellArray  // output cell (array to pass by reference)
            );

            this.applynormal = collided;
        }
        
        this.position.copy(newpos); // newpos is now the collision-adjusted position

        if (this.applynormal) {
            // Original Java code applies normal impulse here.
            // This means bouncing off the wall.
            // v -= normal * (v.normal()) * (1+bouncefactor)
            // For now, just stop movement into wall. A proper bounce needs bouncefactor.
            // A simple way to stop penetration: project velocity onto plane of collision.
            // Or, more simply, if a collision happened, the crossLabyrinth already adjusted newpos.
            // The impulse might need adjustment based on normal.
            // current_impulse_along_normal = this.velocity.dot(this.normal);
            // this.velocity.addScaledVector(this.normal, -current_impulse_along_normal * (1 + bounceFactor) );
            // For now, let's assume crossLabyrinth handles the position adjustment correctly
            // and we might need to adjust velocity/impulse later if we want bounce.
        }

        this.calcRotation(); // Rotates based on this.angle (delta)
        this.applyFriction(); // Apply friction to this.impulse

        this.m_cell = endCellArray[0] ? endCellArray[0] : this.updateCurrentCell();
    }
}
